src/mrack/providers/beaker.py

```diff
@@ -0,0 +1,334 @@
+        bkr_job_xml = self.hub.taskactions.to_xml(beaker_id).encode("utf8")
+
+        resources = []
+        for recipe in eTree.fromstring(bkr_job_xml).iter("recipe"):
+            resources.append(
+                {
+                    "family": recipe.get("family"),
+                    "distro": recipe.get("distro"),
+                    "arch": recipe.get("arch"),
+                    "variant": recipe.get("variant"),
+                    "system": recipe.get("system"),
+                    "status": recipe.get("status"),
+                    "result": recipe.get("result"),
+                    "rid": recipe.get("id"),
+                    "id": recipe.get("job_id"),
+                }
+            )
+
+        return resources[0] if len(resources) == 1 else []
+
+    async def wait_till_provisioned(
+        self, bkr_id_name, timeout=None, poll_sleep=None, max_attempts=None
+    ):
+        """Wait for Beaker provisioning result."""
+        beaker_id, req_hname = bkr_id_name
+
+        if not poll_sleep:
+            poll_sleep = self.poll_sleep
+        if not timeout:
+            timeout = self.timeout
+        if not max_attempts:
+            max_attempts = self.max_attempts
+
+        resource = {}
+        attempts = 0
+        prev_status = ""
+
+        while attempts < self.max_attempts:
+            attempts += 1
+            resource = self._get_recipe_info(beaker_id)
+            status = resource["status"]
+
+            if prev_status != status:
+                logger.info(
+                    f"Job {beaker_id} has changed status ("
+                    f"{prev_status} -> {status})"
+                )
+                prev_status = status
+
+            if STATUS_MAP.get(status) == STATUS_ACTIVE:
+                break
+            elif STATUS_MAP.get(status) == STATUS_PROVISIONING:
+                await asyncio.sleep(self.poll_sleep)
+            elif STATUS_MAP.get(status) in [STATUS_ERROR, STATUS_DELETED]:
+                logger.warning(
+                    f"Job {beaker_id} has errored with status "
+                    f"{status} with result {resource['result']}"
+                )
+            else:
+                logger.error(
+                    f"Job {beaker_id} has swithced to status "
+                    f"{status} with result {resource['result']}"
+                )
+
+        resource.update({"JobID": beaker_id, "hname": req_hname})
+        return resource
+
+    async def provision_hosts(self, hosts):
+        """Provision hosts based on list of host requirements.
+
+        Issues provisioning and waits for it succeed. Raises exception if any of
+        the servers was not successfully provisioned. If that happens it issues deletion
+        of all already provisioned resources.
+
+        Return list of information about provisioned servers.
+        """
+        started = datetime.now()
+
+        count = len(hosts)
+        logger.info(f"Issuing provisioning of {count} hosts")
```

Comment:
A Q for future thoughs, should we log messages like this with some provider prefix? So that we can see in log e.g.:

```
OpenStack: Issuing provisioning of 3 hosts
Beaker: Issuing provisioning of 2 hosts
```

Instead of:
```
Issuing provisioning of 3 hosts
Issuing provisioning of 2 hosts
```

Reply:
Yes that would be nice!

Reply:
We can use filters for that (https://docs.python.org/3/howto/logging-cookbook.html#filters-contextual) in case we don't want to print that explicitly.

Reply:
Let us go this way